### ISSUE TITLE
BearSSL appears to drop the last TLS frame sometimes

### ISSUE THREAD

**Issue by dfandrich** [curl/curl#11326]
The curl test suite frequently fails test 1632 in the BearSSL CI builds with this failure message:
```
test 1632...[FTP through HTTPS-proxy, with connection re-use]

 1632: protocol FAILED:
--- log/check-expected	2023-06-14 08:18:16.353600407 +0000
+++ log/check-generated	2023-06-14 08:18:16.353600407 +0000
@@ -8,4 +8,3 @@
 EPSV[CR][LF]
 SIZE 16320002[CR][LF]
 RETR 16320002[CR][LF]
-QUIT[CR][LF]
```
For example, see this job: https://github.com/curl/curl/actions/runs/5264314339/jobs/9515446230

This is an FTP-over-HTTPS proxy test, and `QUIT` is the last message sent by the client to the server before closing the connection. Since this final message is not being logged by the server, it appears to be becoming lost before it arrives there. My best guess is that this is because  the last message is not being flushed out by curl to the network stack before exiting. However, it's possible that it's a problem in the proxy server instead, but this seems unlikely since it only happens in BearSSL builds in the CI system.

@michaelforney, any ideas?

**Comment by calvin2021y** [comment]
Not sure this related into CURL drop frame problem,  I use curl 8.1.2 with 11247 patch.  get problem from ios arm64 and android arm32.  

error come from this line:
```diff
diff --git a/lib/vtls/wolfssl.c b/lib/vtls/wolfssl.c
index 2928728..6938cb2 100644
--- a/lib/vtls/wolfssl.c
+++ b/lib/vtls/wolfssl.c
@@ -788,7 +788,7 @@ wolfssl_connect_step2(struct Curl_cfilter *cf, struct Curl_easy *data)
     else if(ASN_NO_SIGNER_E == detail) {
       if(conn_config->verifypeer) {
         failf(data, " CA signer not available for verification");
-        return CURLE_SSL_CACERT_BADFILE;
+        return CURLE_SSL_CACERT_BADFILE;
       }
       else {
         /* Just continue with a warning if no strict certificate
```

only RSA domain has problem, the tiny ECC certs domain  has no problem.

https://github.com/wolfSSL/wolfssl/issues/6524

**Comment by calvin2021y** [comment]
try with `CURLOPT_HTTP_VERSION` , `CURL_HTTP_VERSION_1_1`, problem still exits.

**Comment by dfandrich** [comment]
I'm not sure what that problem has to do with this issue. It's not on BearSSL, it doesn't appear to have anything to do with dropping frames. Can you explain?

**Comment by dfandrich** [comment]
Looks like you moved that to #11361

**Comment by michaelforney** [comment]
I'm not very familiar with ftp and http proxy details, but I am able to reproduce the test failure. I did some poking around, but I'm still not sure what's going on.

It seems that test 1632 always passes when I run with valgrind disabled, but sometimes failed as you describe with valgrind. Comparing the logs, I see the following in the failed `http2_server.log`:

```
 08:17:45.308530 [CTRL] READ 15 bytes from client
 08:17:45.308583 [CTRL] READ "RETR 16320002%0d%0a"
 08:17:45.308640 [CTRL] SENT 15 bytes to server
 08:17:45.308666 [CTRL] SENT "RETR 16320002%0d%0a"
 08:18:16.342163 CONNECT proxy timeout after 31 idle seconds
 08:18:16.342295 [CTRL] CLOSING server socket (cleanup)
 08:18:16.342436 [CTRL] CLOSING client socket (cleanup)
 08:18:16.342516 [CTRL] ABORTING
```

On a successful run, it looks something like this:

```
14:05:14.142900 [CTRL] READ 15 bytes from client
14:05:14.142923 [CTRL] READ "RETR 16320002%0d%0a"
14:05:14.142949 [CTRL] SENT 15 bytes to server
14:05:14.142960 [CTRL] SENT "RETR 16320002%0d%0a"
14:05:14.143522 [CTRL] READ 53 bytes from server
14:05:14.143548 [CTRL] READ "150 Binary data connection for 1632 () (323 bytes).%0d%0a"
14:05:14.143576 [CTRL] SENT 53 bytes to client
14:05:14.143587 [CTRL] SENT "150 Binary data connection for 1632 () (323 bytes).%0d%0a"
<snip>
14:05:14.644848 [DATA] CLOSING client socket
14:05:14.644909 [DATA] CLOSING server socket
14:05:14.644927 [DATA] ENDING
14:05:14.644979 [CTRL] SENT 28 bytes to client
14:05:14.644997 [CTRL] SENT "226 File transfer complete%0d%0a"
14:05:14.645882 [CTRL] READ 6 bytes from client
14:05:14.645910 [CTRL] READ "QUIT%0d%0a"
14:05:14.645942 [CTRL] SENT 6 bytes to server
14:05:14.645957 [CTRL] SENT "QUIT%0d%0a"
14:05:14.646457 [CTRL] READ 18 bytes from server
14:05:14.646493 [CTRL] READ "221 bye bye baby%0d%0a"
14:05:14.646532 [CTRL] SENT 18 bytes to client
14:05:14.646548 [CTRL] SENT "221 bye bye baby%0d%0a"
14:05:14.646799 [CTRL] got 0, STOP READING client
14:05:14.646829 [CTRL] DISABLED WRITING server
14:05:14.897167 [CTRL] got 0, STOP READING server
14:05:14.897218 [CTRL] DISABLED WRITING client
14:05:15.147564 [CTRL] CLOSING client socket
14:05:15.147621 [CTRL] CLOSING server socket
14:05:15.147639 [CTRL] ENDING
```

So it seems in the failure case that the HTTP2 proxy is receiving `RETR 16320002\r\n` from curl and relaying it to the FTP server (and looking at `ftp_sockctrl.log`, the FTP server does see this), but after this, the FTP server seems to hang and sends no data back to curl. After 31 seconds, there is a timeout and the connection is closed.

Looking at `ftp_sockctrl.log`, I see

```
 08:17:45.308674 < 15 bytes data, client => server
 08:17:45.308722 'RETR 16320002\r\n'
 08:18:16.342542 ====> Client disconnect
```

On a successful run, it looks like this

```
14:05:14.142957 < 15 bytes data, client => server
14:05:14.142972 'RETR 16320002\r\n'
14:05:14.143442 Received DATA (on stdin)
14:05:14.143459 > 53 bytes data, server => client
14:05:14.143471 '150 Binary data connection for 1632 () (323 bytes).\r\n'
14:05:14.144300 Received DATA (on stdin)
14:05:14.144341 > 28 bytes data, server => client
14:05:14.144366 '226 File transfer complete\r\n'
14:05:14.645997 < 6 bytes data, client => server
14:05:14.646057 'QUIT\r\n'
14:05:14.646307 Received DATA (on stdin)
14:05:14.646344 > 18 bytes data, server => client
14:05:14.646373 '221 bye bye baby\r\n'
14:05:14.646898 ====> Client disconnect
```

So why does the FTP server not send back the data after the `RETR` command? Is it waiting for something on the client end?

**Comment by bagder** [comment]
Does this problem still occur? I don't recall seeing this trigger in the CI builds?

**Comment by michaelforney** [comment]
I wasn't able to reproduce this on git master, but as a sanity check, I checked out a commit from June 2023 and confirmed that I could still reproduce it on that commit. I bisected it to 691c32f6f5fde67d858ffa25dc5a194dffdf21ab, which perhaps affected the timing of the TLS connection?

To test this theory, on current master, I added a `sleep(2)` right before `br_ssl_client_init_full` and sure enough, I saw the failure again. It seems that the recent performance improvements to the bearssl backend (691c32f6f5fde67d858ffa25dc5a194dffdf21ab, 1bccee76c8211da3233f08444abacd7cc00d33d4) may have kept the test passing in CI, but I think the theoretical issue is still there.

I'll do some more investigation tomorrow now that I can reproduce it reliably.

**Comment by michaelforney** [comment]
Ok, I think I've figured out what's going on. The actual important log is in `ftp_server.log`:

```
03:43:19.630791 ====> Passive DATA channel requested by client
03:43:19.630842 DATA sockfilt for passive data channel starting...
03:43:19.632764 DATA sockfilt for passive data channel started (pid 11524)
03:43:19.633040 DATA sockfilt for passive data channel listens on port 43713
03:43:19.633175 > "229 Entering Passive Mode (|||43713|)[LF]"
03:43:19.633238 Client has been notified that DATA conn will be accepted on port 43713
03:43:21.633450 FTP server timed out awaiting data connection on port 43713
03:43:21.633560 accept failed or connection not even attempted
03:43:21.633600 DATA sockfilt killed now
03:43:21.634076 DATA sockfilt not running
```

The FTP server has a 2 second timeout for a connection to its DATA port. It seems that a combination of BearSSL reading CAs unnecessarily, slow TLS handshakes due to unoptimal polling, valgrind slowdown, and going through the HTTP proxy were enough for that timeout to be exceeded sometimes. This explains why I needed `sleep(2)` to trigger it every time (`sleep(1)` wasn't enough).

So I'm going to call this fixed by 691c32f6f5fde67d858ffa25dc5a194dffdf21ab and 1bccee76c8211da3233f08444abacd7cc00d33d4.

**Comment by bagder** [comment]
Presumably fixed!
